File: api/app/concepts/effect/prevent_actions.rb
class Effect::PreventActions < Effect::Base
  delegate :affected_players, :affected_tiles, to: :area_of_effect

  def sort_order
    OVERRIDE_ORDER
  end

  def results
    compute_results(EffectResult::AttemptPreventActions, tiles: affected_tiles) +
      target_players.flat_map { |player| compute_results(EffectResult::PreventActions, player: player, duration_type: card.duration_type, duration: card.duration) }
  end

  private

  def target_players
    affected_players.select { |p| p != player }
  end

  def area_of_effect
    AreaOfEffect.new(players: game.players, tiles: game.tiles, ranges: card.card_ranges)
  end
end

File: common/src/gameplay/turnResults/preventActions.ts
import { List, Map } from "immutable"
import modifiedResults from "./helpers/modifiedResults"
import { ActionResult } from "./resultTypes"

import { BoardTile } from "../state/board"
import { Card, PreventActionsEffect } from "../state/card"
import { Duration } from "../state/duration"
import { GameState } from "../state/gameState"
import { affectedPlayers, affectedTiles } from "../state/helpers/range"
import { Player } from "../state/player"

export function calculatePreventActionsResults(
  playedCards: Map<Player, Card>,
  gameState: GameState
): List<ActionResult> {
  return playedCards.flatMap((card, player) =>
    resultsOf(effectsFrom(card, player), gameState)
  ) as List<ActionResult>
}

function effectsFrom(card: Card, player: Player) {
  return card.effects
    .filter(effect => effect.type === "preventActions")
    .map(effect => [effect as PreventActionsEffect, player]) as List<[PreventActionsEffect, Player]>
}

function resultsOf(effects: List<[PreventActionsEffect, Player]>, gameState: GameState) {
  return effects.flatMap(([effect, player]) => effectResults(effect, player, gameState)) as List<
    ActionResult
  >
}

function effectResults(effect: PreventActionsEffect, player: Player, gameState: GameState) {
  const tiles = affectedTiles(effect.ranges, player.position, gameState.board)
  const players = affectedPlayers(tiles, gameState)

  return List([attemptPreventActionsResult(tiles)]).concat(
    preventPlayerActionsResults(effect.duration, players)
  )
}

function attemptPreventActionsResult(tiles: List<BoardTile>) {
  return { type: "attemptPreventActions", tiles }
}

function preventPlayerActionsResults(duration: Duration, players: List<Player>) {
  return players.flatMap(player =>
    modifiedResults({ type: "preventActions", duration, player }, player)
  )
}
